File: test/InterpreterTest.cpp
#include <gtest/gtest.h>
#include "../src/Interpreter.h"
#include "../src/SceneStatement.h"
#include "../src/Knowledge.h"

TEST(InterpreterTests, interpreterTest1) {

    SceneModel model;
    Knowledge knowledge;

    std::vector<std::shared_ptr<SceneStatement> > result = interpret("Create a box.", model, knowledge, false);

    std::shared_ptr<CreateEntityRule> stmt;

    ASSERT_TRUE(stmt = std::dynamic_pointer_cast<CreateEntityRule>(result[0]));

    ASSERT_EQ(1, stmt->number);

    ASSERT_EQ("box", stmt->what);

}

TEST(InterpreterTests, interpreterTest2) {

    SceneModel model;
    Knowledge knowledge;

    std::vector<std::shared_ptr<SceneStatement> > result = interpret("Add a palm tree or two.", model, knowledge, false);

    std::shared_ptr<CreateEntityRule> stmt;

    ASSERT_TRUE(stmt = std::dynamic_pointer_cast<CreateEntityRule>(result[0]));

    ASSERT_LE(1, stmt->number);
    ASSERT_GE(2, stmt->number);

    ASSERT_EQ("tree", stmt->what);

}

TEST(InterpreterTests, createWithAnd) {

    SceneModel model;
    Knowledge knowledge;

    std::vector<std::shared_ptr<SceneStatement> > result = interpret("Add a box and a cylinder", model, knowledge, false);

    ASSERT_EQ(2, result.size());

    std::shared_ptr<CreateEntityRule> stmt;

    ASSERT_TRUE(stmt = std::dynamic_pointer_cast<CreateEntityRule>(result[0]));

    ASSERT_EQ(1, stmt->number);

    ASSERT_EQ("box", stmt->what);

    ASSERT_TRUE(stmt = std::dynamic_pointer_cast<CreateEntityRule>(result[1]));

    ASSERT_EQ(1, stmt->number);

    ASSERT_EQ("cylinder", stmt->what);

}

File: test/InterpreterTest.java
import org.junit.Assert;
import org.junit.Test;

import java.util.List;

public class InterpreterTest {

    Interpreter getInterpreter(CompositeModel model) {
        Knowledge knowledge = Knowledge.knowledgeWithPrimitives();

        return new Interpreter(new Resolver(null, knowledge) {
            @Override
            void askUserForObject(String name) {
                // Should not ask for clarification
                Assert.fail();
            }
        });
    }

    @Test
    public void interpreterTest1() {

        CompositeModel scene = new CompositeModel("Scene");
        List<SceneCommand> result = getInterpreter(scene).interpret("Create a cube.", scene);

        Assert.assertTrue(result.get(0) instanceof CreateEntityRule);

        CreateEntityRule stmt = (CreateEntityRule) result.get(0);

        Assert.assertEquals(1, stmt.number);

        Assert.assertEquals("Cube", stmt.what.getName());

    }

    @Test
    public void interpreterTest2() {

        CompositeModel scene = new CompositeModel("Scene");

        List<SceneCommand> result = getInterpreter(scene).interpret("Add a palm tree or two.", scene);

        Assert.assertTrue(result.get(0) instanceof CreateEntityRule);

        CreateEntityRule stmt = (CreateEntityRule) result.get(0);

        Assert.assertTrue(1 <= stmt.number);
        Assert.assertTrue(2 >= stmt.number);

        Assert.assertEquals("tree", stmt.what.getName());

    }

    @Test
    public void interpreterTestNumeric() {

        CompositeModel scene = new CompositeModel("Scene");

        List<SceneCommand> result = getInterpreter(scene).interpret("50 spheres", scene);

        Assert.assertTrue(result.get(0) instanceof CreateEntityRule);

        CreateEntityRule stmt = (CreateEntityRule) result.get(0);

        Assert.assertEquals(50, stmt.number);

        Assert.assertEquals("Sphere", stmt.what.getName());

    }

    @Test
    public void createWithAnd() {

        CompositeModel scene = new CompositeModel("Scene");

        List<SceneCommand> result = getInterpreter(scene).interpret("Add a cube and a cylinder", scene);

        Assert.assertEquals(2, result.size());


        CreateEntityRule stmt = (CreateEntityRule) result.get(0);

        Assert.assertTrue(result.get(0) instanceof CreateEntityRule);

        Assert.assertEquals(1, stmt.number);

        Assert.assertEquals("Cube", stmt.what.getName());


        Assert.assertTrue(result.get(1) instanceof CreateEntityRule);

        stmt = (CreateEntityRule) result.get(1);

        Assert.assertEquals(1, stmt.number);

        Assert.assertEquals("Cylinder", stmt.what.getName());

    }
}